bee-network/bee-autopeering/src/peering/manager.rs

```diff
@@ -0,0 +1,774 @@
+            log::debug!("Failed to add neighbor to outbound neighborhood after successful peering request");
+        }
+    } else {
+        log::debug!("Peering by {} denied.", ctx.peer_id);
+    }
+
+    // Send the response notification.
+    if let Some(tx) = peer_reqval.response_tx {
+        tx.send(ctx.msg_bytes.to_vec()).expect("error sending response signal");
+    }
+}
+
+fn handle_drop_request<V: NeighborValidator>(
+    _drop_req: DropPeeringRequest,
+    ctx: RecvContext,
+    nb_filter: &NeighborFilter<V>,
+) {
+    log::trace!("Handling drop request.");
+
+    let mut removed_nb = ctx.inbound_nbh.remove_neighbor(ctx.peer_id);
+
+    if let Some(nb) = ctx.outbound_nbh.remove_neighbor(ctx.peer_id) {
+        removed_nb.replace(nb);
+
+        nb_filter.add(*ctx.peer_id);
+
+        // TODO: trigger immediate outbound neighborhood update; currently we wait for the next interval
+    }
+
+    if removed_nb.is_some() {
+        send_drop_peering_request_to_addr(
+            ctx.peer_addr,
+            *ctx.peer_id,
+            ctx.server_tx,
+            ctx.event_tx,
+            ctx.inbound_nbh,
+            ctx.outbound_nbh,
+        );
+    }
+}
+
+///////////////////////////////////////////////////////////////////////////////////////////////////////////
+// SENDING
+///////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+/// Initiates a peering request.
+///
+/// This function is blocking, but at most for `RESPONSE_TIMEOUT` seconds.
+pub(crate) async fn begin_peering(
+    peer_id: &PeerId,
+    active_peers: &ActivePeersList,
+    request_mngr: &RequestManager,
+    server_tx: &ServerTx,
+    local: &Local,
+) -> Option<Status> {
+    let (response_tx, response_rx) = request::response_chan();
+
+    send_peering_request_to_peer(peer_id, active_peers, request_mngr, server_tx, Some(response_tx), local);
+
+    match tokio::time::timeout(RESPONSE_TIMEOUT, response_rx).await {
+        Ok(Ok(bytes)) => {
+            let peer_res = PeeringResponse::from_protobuf(&bytes).expect("error decoding peering response");
+            Some(peer_res.status())
+        }
+        Ok(Err(e)) => {
+            log::debug!("Response signal error: {}", e);
+            None
+        }
+        Err(e) => {
+            log::debug!("Peering response timeout: {}", e);
+
+            // The response didn't arrive in time => remove the request.
+            let _ = request_mngr.write().pull::<PeeringRequest>(peer_id);
+
+            None
+        }
+    }
+}
+
+/// Sends a peering request to a peer.
+pub(crate) fn send_peering_request_to_peer(
+    peer_id: &PeerId,
+    active_peers: &ActivePeersList,
+    request_mngr: &RequestManager,
+    server_tx: &ServerTx,
+    response_tx: Option<ResponseTx>,
+    local: &Local,
+) {
+    // TEMP: guard and drop
+    let guard = active_peers.read();
+    if let Some(active_peer) = active_peers.read().find(peer_id) {
+        drop(guard);
```

Comment:
Not sure what's happening here, why do you take the guard two times ?

Reply:
I can't quite remember what happened there. Probably it looks weird now because I refactored something. Anyway....it's fixed now. Thanks for drawing attention to this. It seems like I knew this couldn't possibly make it into the release, that's why the TEMP, but I should have better explained why I temporarily needed that. :wink: